### DISCUSSION TITLE
Termination redux

### DISCUSSION THREAD

**Discussion by mturilli** [radical-cybertools/radical.pilot#2411] [Ideas]
Due to a long list of Python bugs, termination in RP is functional but over-engineered. One of the critical bugs was about  destructors/garbage collector behavior in a threading context. Luckily, this bug has been fixed and back-ported to Python 3.6. @andre-merzky tested it and it worked. We can now simplify termination in RP and, as a consequence, eliminate (part of?) the heartbeat.

**Comment by mturilli** [comment]
Roadmap: this should be done after env isolation and before moving the tracer (and possibly other sub-components) out of RP into stand-alone components.

**Comment by andre-merzky** [comment]
We should be able to revert from heartbeat messages to process monitoring: parents watch child processes (components) and react on their disappearance; child processes watch parents and initiate termination on their disappearance.  This implies one watcher thread per process.  Component destructors should actively kill children on termination.  All threads should be daemon threads.
